enhancements/cloud-integration/openstack/failure-domains.md

```diff
@@ -0,0 +1,720 @@
+
+We will create a new `OpenStackFailureDomain` type defining a set of Machine
+properties which together ensure the Machine is deployed to a specific,
+consistent failure domain. These properties will be merged into the Machine
+definition by the machine provisioner, which may be either the installer or the
+machine controller.
+
+```go
+type OpenStackFailureDomain struct {
+    // ComputeAvailabilityZone is the name of a valid nova availability zone
+    // +optional
+    ComputeAvailabilityZone string `json:"computeAvailabilityZone,omitempty"`
+
+    // StorageAvailabilityZone is the name of a valid cinder availability zone
+    // +optional
+    StorageAvailabilityZone string `json:"storageAvailabilityZone,omitempty"`
+
+    // Ports defines a set of ports and their attached networks. These will be prepended to any another ports attached to the server.
+    // +optional
+    Ports []PortOpts `json:"ports,omitempty"`
+}
+```
+
+This same struct will be referenced:
+* In the installer's InstallConfig for the initial creation of control plane servers and the creation of relevant manifests.
+* In OpenStackProviderSpec for the creation of a machine using a specific failure domain.
+* In CPMS, for the managed balancing of control plane machines across failure domains.
+
+Leaving any of these fields as their zero value will cause it to be ignored when provisioning the machine. This means a failure domain can define any combination of these fields.
+
+`ComputeAvailabilityZone` will set the Nova availability zone of the created server. It will be an error to specify [`AvailabilityZone`](https://github.com/openshift/api/blob/54592eea55395af31e82fa9a605eb45523115454/machine/v1alpha1/types_openstack.go#L67-L69) on a Machine with a failure domain that specifies `ComputeAvailabilityZone`.
+
+`StorageAvailabilityZone` will set the Cinder availability zone of any volume attached during initial machine creation. This is currently limited to a root volume, but may be extended to include additional volumes in the future. It will be an error to specify [`Zone`](https://github.com/openshift/api/blob/54592eea55395af31e82fa9a605eb45523115454/machine/v1alpha1/types_openstack.go#L358-L360) on a volume attached to a Machine with a failure domain that specifies `StorageAvailabilityZone`.
+
+`Ports` will be used to define a set of port to be attached to all Machines in a failure domain. They use the existing [`PortOpts`](https://github.com/openshift/api/blob/54592eea55395af31e82fa9a605eb45523115454/machine/v1alpha1/types_openstack.go#L283-L333) struct to define ports with all currently supported properties and attachments with the same semantics. Note specifically that the network or subnet a port will attach to must already exist. Unlike `ComputeAvailabilityZone` and `StorageAvailabilityZone` it will be allowed to specify this field on both the failure domain and the Machine. If both are defined they will be merged by putting the failure domain ports first.
```

Comment:
Since we know `PortOpts` is problematic, why not introduce this new field with a `PortOptsNew` which has a better design? 

For CPMS, you will introduce a new API, the failure domain for openstack will be within the remit of the CPMS and machine v1 APIs, and so will have to go through thorough review.

It must be convertible to the existing `PortOpts`, but need not be identical. I would encourage you to think about this new failure domain ports using the same design